db/keyval/consul/plugin.go

```diff
@@ -88,50 +93,54 @@ func ConfigToClient(cfg *Config) (*api.Config, error) {
 
 // Init initializes Consul plugin.
 func (plugin *Plugin) Init() (err error) {
-	cfg, err := plugin.getConfig()
-	if err != nil || plugin.disabled {
-		return err
-	}
-	clientCfg, err := ConfigToClient(cfg)
-	if err != nil {
-		return err
-	}
-	plugin.client, err = NewClient(clientCfg)
-	if err != nil {
-		plugin.Log.Errorf("Err: %v", err)
-		return err
-	}
-	plugin.reconnectResync = cfg.ReconnectResync
-	plugin.protoWrapper = kvproto.NewProtoWrapperWithSerializer(plugin.client, &keyval.SerializerJSON{})
-
-	// Register for providing status reports (polling mode).
-	if plugin.StatusCheck != nil {
-		plugin.StatusCheck.Register(core.PluginName(plugin.PluginName), func() (statuscheck.PluginState, error) {
-			_, _, _, err := plugin.client.GetValue(healthCheckProbeKey)
-			if err == nil {
-				if plugin.reconnectResync && plugin.lastConnErr != nil {
-					plugin.Log.Info("Starting resync after Consul reconnect")
-					if plugin.Resync != nil {
-						plugin.Resync.DoResync()
-						plugin.lastConnErr = nil
-					} else {
-						plugin.Log.Warn("Expected resync after Consul reconnect could not start beacuse of missing Resync plugin")
+	plugin.initOnce.Do(func() {
+		cfg, err := plugin.getConfig()
+		if err != nil || plugin.disabled {
+			return
+		}
```

Comment:
You've got a pretty gnarly bug here that means this will _never_ return an error. When you do `err := ...` you're shadowing the `err` variable you've defined at the top of the function. Shadowing means you're creating a variable with the same name, that prevents you from directly accessing the original variable.

Because you're shadowing the `err` variable defined in the function header, it means you aren't mutating the outer `err` variable. So when you do a bare `return` you're *_always_* going to return `nil`. Do you want to always return `nil`, or the error from `plugin.getConfig()`?

There are probably at least 10 other examples of that type of shadowing in this PR alone, and it seems there are more lurking in the codebase / other PRs.